apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/AbstractSamhsaMatcher.java

```diff
@@ -0,0 +1,374 @@
+       */
+      throw new BadCodeMonkeyException(e);
+    }
+  }
+
+  /**
+   * Checks that for the specified {@link CodeableConcept}, the Codings (if any) within, contain a
+   * blacklisted MHSA procedure code. If any of the systemms within th {@link CodeableConcept} are
+   * not known/expected, returns {@code true} and assuems the system is SAMHSA as a safety fallback.
+   *
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains any
+   *     {@link Coding}s that match any of the {@link #cptCodes} or has unknown coding systems,
+   *     <code>false</code> otherwise.
+   */
+  protected boolean containsSamhsaProcedureCode(CodeableConcept procedureConcept) {
+    return !procedureConcept.getCoding().isEmpty()
+        && (hasHcpcsSystemAndSamhsaCptCode(procedureConcept)
+            || !containsOnlyKnownSystems(procedureConcept));
+  }
+
+  /**
+   * Checks that for a {@link CodeableConcept} the {@link Coding}s contain a HCPCS system and at
+   * least one blacklisted CPT code.
+   *
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains a HCPCS
+   *     system as well as a CPT samhsa coding, (matched with {@link #cptCodes}), <code>false</code>
+   *     otherwise.
+   */
+  @VisibleForTesting
+  boolean hasHcpcsSystemAndSamhsaCptCode(CodeableConcept procedureConcept) {
+    // CPT codes are a subset of HCPCS codes, but they are the only ones we blacklist
+
+    // If there is no HCPCS system, it may be a DME claim, which should return false for SAMHSA
+    return procedureConcept.getCoding().stream()
+            .anyMatch(code -> TransformerConstants.CODING_SYSTEM_HCPCS.equals(code.getSystem()))
+        && procedureConcept.getCoding().stream().anyMatch(this::isSamhsaCptCode);
+  }
+
+  /**
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains at least
+   *     one {@link Coding} and only contains {@link Coding}s that have known coding systems <code>
+   *     false</code> otherwise
+   */
+  @VisibleForTesting
+  boolean containsOnlyKnownSystems(CodeableConcept procedureConcept) {
+    Set<String> codingSystems =
+        procedureConcept.getCoding().stream().map(Coding::getSystem).collect(Collectors.toSet());
+
+    String hcpcsCdSystem = CCWUtils.calculateVariableReferenceUrl(CcwCodebookVariable.HCPCS_CD);
+
+    // Valid system url for productOrService coding
+    final Set<String> hcpcsSystem = Set.of(TransformerConstants.CODING_SYSTEM_HCPCS);
+
+    // Additional valid coding system URL for backwards-compatibility
+    // See: https://jira.cms.gov/browse/BFD-1345
+    Set<String> backwardsCompatibleHcpcsSystems =
+        Set.of(TransformerConstants.CODING_SYSTEM_HCPCS, hcpcsCdSystem);
```

Comment:
So the `https://bluebutton.cms.gov/resources/variables/hcpcs_cd` system should be considered an unknown system for v1, and thus assumed to be SAMHSA?

Reply:
Yes, thats how it is currently implemented prior to this PR; I would say a refactor should retain the same implementation unless someone has a compelling reason why we would allow for HCPCS_CD for v1.

As a followup to this, there _should_ be a unit test written for v1 to ensure that HCPCS_CD was/is assumed to be SAMHSA for v1, and not for v2, which should then subsequently continue passing after this adjustment. This would be the best way to ensure that nothing broke as a result of this and we retain the same functionality.